feat: documentation for manual installation of deb archive (#202)
canonical/landscape-documentation: docs/how-to-guides/landscape-installation-and-set-up/debarchive-repository-management.md

```

```{note}
When the configuration shim reads `/etc/landscape/service.conf`, the values from that file take precedence for the fields it maps (database name, host, user, password, and secrets). Values set via `snap set` for those fields are only used if `service.conf` is not available.
```

### Restart the service after configuration changes

The snap automatically restarts when configuration changes are applied via `snap set`. If you edited `service.conf` directly, restart the service manually:

```bash
sudo snap restart landscape-debarchive
```

## Configure the reverse proxy

The Deb Archive gateway must be exposed at the `/debarchive` path on your Landscape Server's external URL. This requires updating your reverse proxy to forward requests to the Deb Archive service while stripping the `/debarchive` prefix.

### Apache

If you followed the {ref}`manual installation guide <how-to-heading-manual-install-configure-web-server>`, your Landscape Server uses Apache. Add the following rules to the `<VirtualHost *:443>` block in `/etc/apache2/sites-available/landscape.conf` (or the appropriate configuration file for your setup).

> Add these lines **before** the final catch-all `RewriteRule` at the bottom of the block (the line that rewrites to `localhost:8080`):

Reviewer: nit: could specify what this line looks like in a bit more detail

Author: I'll add a bit more detail to the parenthetical.

Add an ACL and `use_backend` directive in the existing `frontend` section:

```haproxy
    acl is_debarchive path_beg /debarchive
    use_backend debarchive if is_debarchive
```

Add a new backend section:

```haproxy
backend debarchive
    http-request set-path %[path,regsub(^/debarchive,/)]
    server debarchive 127.0.0.1:8100 check
```

```{note}
If the Deb Archive service runs on a different machine from HAProxy, replace `127.0.0.1` with the appropriate IP address or hostname.
```

Then reload HAProxy:

```bash
sudo systemctl reload haproxy
```

## Verify the installation

### Check the service status

Confirm the snap service is running:

```bash
sudo snap services landscape-debarchive
```

The output should show the `debarchive` service as **active**:

```text
Service                              Startup  Current  Notes
landscape-debarchive.debarchive      enabled  active   -
```

### Check the health endpoint

Send a health check request through the reverse proxy. Replace `<LANDSCAPE_FQDN>` with the FQDN of your Landscape Server:

```bash

> curl -sk -o /dev/null -w "%{http_code}" https://<LANDSCAPE_FQDN>/debarchive/v1/mirrors

Reviewer: nit: I prefer having things like <LANDSCAPE_FQDN> as env vars so they can be copy pasted easier. Not sure if we have guidance against doing so but I think it's a little more helpful for the user.

Author: Fine idea. I'll adjust it.

```

> A response of `401` (unauthorized) confirms the Deb Archive service is reachable through the reverse proxy.

Reviewer: Maybe worth mentioning debarchive auth goes through Landscape in case a user want to confirm an endpoint with a nonerror response.

Author: Yes, I'll mention that.

You can also test directly against the service (bypassing the proxy) to isolate connectivity issues:

```bash
curl -s -o /dev/null -w "%{http_code}" http://localhost:8100/v1/mirrors
```

### Verify in the Landscape web portal

1. Log in to the Landscape web portal at `https://<LANDSCAPE_FQDN>`
2. Navigate to the **repository management** page
3. Confirm that you can create or add a new repository mirror

If the repository management page loads and allows you to begin adding mirrors, the Deb Archive service is fully operational.

## Troubleshooting

### Service fails to start

Check the snap logs for error details:

```bash
sudo snap logs landscape-debarchive -n 50
```

Common issues include:

> - **Database connection errors**: Verify the database host, port, user, and password. Ensure the `landscape-debarchive` database exists and the configured user has access.

Reviewer: Throughout the guide the database name is referred to as `landscape-standalone-debarchive`. Not sure if you mean to use the same name here or not.

Author: I did. Thanks for catching it.

### Apache

If you followed the {ref}`manual installation guide <how-to-heading-manual-install-configure-web-server>`, your Landscape Server uses Apache. Add the following rules to the `<VirtualHost *:443>` block in `/etc/apache2/sites-available/landscape.conf` (or the appropriate configuration file for your setup).

Add these lines **before** the final catch-all `RewriteRule` at the bottom of the block (the line that starts with `RewriteRule ^/(.*) http://localhost:8080/...`):

```apache
    # Landscape Deb Archive
    RewriteRule ^/debarchive$ /debarchive/ [R=permanent]
    RewriteRule ^/debarchive/(.*) http://localhost:8100/$1 [P,L]
```

Then reload Apache:

```bash
sudo systemctl reload apache2
```

### HAProxy

If your deployment uses HAProxy as the reverse proxy, add a backend and routing rule for the Deb Archive service. Edit your HAProxy configuration (typically `/etc/haproxy/haproxy.cfg`):

Add an ACL and `use_backend` directive in the existing `frontend` section:

> ```haproxy

Reviewer 2: Sphinx doesn't know what this is - you need to change the code block type (text, bash, ...)

Author: That's because it's a typo. Well, a placeholder I forgot to remove. I don't think haproxy config files are any specific syntax so I'll just leave these unspecified.

    acl is_debarchive path_beg /debarchive
    use_backend debarchive if is_debarchive
```

Add a new backend section:

> ```haproxy

Reviewer 2: Sphinx doesn't know what this is - you need to change the code block type (text, bash, ...)

Deb Archive was introduced in **Landscape 26.04 LTS**.

## Prerequisites

You need one of the following Landscape Server installations:
- A {ref}`quickstart <how-to-quickstart-installation>` installation of `landscape-server-quickstart`
- A {ref}`manual <how-to-manual-installation>` installation of `landscape-server`

For Quickstart installations, you need administrative access to the Landscape Server machine.

For Manual installations, you need:

- Access to the PostgreSQL database server used by Landscape
- Administrative (root) access to the application server

## Install the landscape-debarchive snap

Install the snap on the same machine as the Landscape application server. 

- For Quickstart installations, this is the same machine as your Landscape server.
- For Manual installations, this is the Landscape application server.

```bash

> sudo snap install landscape-debarchive --channel=latest/edge

Reviewer 3: ```suggestion
sudo snap install landscape-debarchive --channel=latest/beta
```

```bash
grep debarchive /etc/landscape/service.conf
```

Example output:

```ini
debarchive = landscape-standalone-debarchive
```

Which should match the name of the database you created in the previous step.

If the entry is missing, manually add the `debarchive = <DATABASE_NAME>` to the `[stores]` section of `/etc/landscape/service.conf`.

```{note}
**Manual installations**: The Deb Archive snap must be able to read `/etc/landscape/service.conf`. If this file isn't on on the machine you're installing Deb Archive on, you'll need to manually copy it to that machine.
```

### (Optional) Override default settings with `snap set`

Quickstart installations use the default configurations and generally don't require additional setup. Skip this section unless you've customized your database configuration.

For Manual installations, if your PostgreSQL server is **not** at the default location (`localhost:5432`), or you need to override any other defaults, use `snap set` to configure the snap directly. 

The available settings and their defaults are:

| Setting | Snap key | Default |
|---|---|---|
| Gateway (HTTP) port | `deb.archive.server.gateway-port` | `8100` |

> | gRPC port | `deb.archive.server.grpc-port` | `8101` |

Reviewer 3: ```suggestion
```

| Server host | `deb.archive.server.host` | `localhost` |
| Database driver | `deb.archive.database.driver` | `pgx` |

> | Database name | `deb.archive.database.name` | `db` |

Reviewer 3: ```suggestion
| Database name | `deb.archive.database.name` | *(empty, read from service.conf)* |
```

> | Database host | `deb.archive.database.host` | `localhost` |

Reviewer 3: ```suggestion
| Database host | `deb.archive.database.host` | *(empty, read from service.conf)* |
```

| Database port | `deb.archive.database.port` | `5432` |
| Database user | `deb.archive.database.user` | *(empty, read from service.conf)* |
| Database password | `deb.archive.database.password` | *(empty, read from service.conf)* |
| Database SSL mode | `deb.archive.database.ssl` | `disable` |

> | Logging level | `deb.archive.logging.level` | `debug` |

Reviewer 3: ```suggestion
| Logging level | `deb.archive.logging.level` | `info` |
```

> | Human-readable logs | `deb.archive.logging.human-readable` | `true` |

Reviewer 3: ```suggestion
| Human-readable logs | `deb.archive.logging.human-readable` | `false` |
```

> | Filesystem storage path | `deb.archive.filesystem-storage-path` | `./filesystem_storage` |

Reviewer 3: ```suggestion
| Filesystem storage path | `deb.archive.filesystem-storage-path` | `$SNAP_COMMON/filesystem_storage` |
```

    RewriteRule ^/debarchive/(.*) http://localhost:8100/$1 [P,L]
```

Then reload Apache:

```bash
sudo systemctl reload apache2
```

### HAProxy (some Manual installations)

If your deployment uses HAProxy, add a routing rule and backend for the Deb Archive service. Edit your HAProxy configuration (typically `/etc/haproxy/haproxy.cfg`):

In the existing `frontend` section, add:

```
    acl is_debarchive path_beg /debarchive
    use_backend debarchive if is_debarchive
```

Add a new backend section:

```
backend debarchive
    http-request set-path %[path,regsub(^/debarchive,/)]
    server debarchive 127.0.0.1:8100 check
```

```{note}
If the Deb Archive service runs on a different machine from HAProxy, replace `127.0.0.1` with the appropriate IP address or hostname.
```

Then reload HAProxy:

```bash
sudo systemctl reload haproxy
```

## Verify the installation

### Check the service status

Confirm the snap service is running:

```bash
sudo snap services landscape-debarchive
```

The output should show the `debarchive` service as **active**:

```text
Service                              Startup  Current  Notes
landscape-debarchive.debarchive      enabled  active   -
```

> ### Check the health endpoint

Reviewer 3: ```suggestion
### Verify Deb Archive is reachable
```

> Send a health check request through the reverse proxy. Replace `$LANDSCAPE_FQDN` with the FQDN of your Landscape Server, or set it as an environment variable:

Reviewer 3: ```suggestion
Send a probe request through the reverse proxy. Replace `$LANDSCAPE_FQDN` with the FQDN of your Landscape Server, or set it as an environment variable:
```

```bash

> curl -sk -o /dev/null -w "%{http_code}" "https://$LANDSCAPE_FQDN/debarchive/v1/mirrors"

Reviewer 3: ```suggestion
curl -sk -o /dev/null -w "%{http_code}" "https://$LANDSCAPE_FQDN/debarchive/v1beta1/mirrors"
```

```

A response of `401` (unauthorized) confirms the Deb Archive service is reachable through the reverse proxy. Deb Archive uses the same authentication as the main Landscape Server API. You should receive a `200` response instead if you include a JWT token in the request via bearer authentication, or include a cookie from a logged-in session in the Landscape web portal.

You can also test directly against the service (bypassing the proxy) to isolate connectivity issues:

```bash

> curl -s -o /dev/null -w "%{http_code}" http://localhost:8100/v1/mirrors

Reviewer 3: ```suggestion
curl -s -o /dev/null -w "%{http_code}" http://localhost:8100/v1beta1/mirrors
```
